src/operator/random/shuffle_op.cc

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2018 by Contributors
+ * \file shuffle_op.cc
+ * \brief Operator to shuffle elements of an NDArray
+ */
+#if (__GNUC__ > 4 && !defined(__clang__major__)) || (__clang_major__ > 4 && __linux__)
+  #define USE_GNU_PARALLEL_SHUFFLE
+#endif
+
+#include <mxnet/operator_util.h>
+#include <algorithm>
+#include <random>
+#include <vector>
+#ifdef USE_GNU_PARALLEL_SHUFFLE
+  #include <parallel/algorithm>
+#endif
+#include "../elemwise_op_common.h"
+
+namespace mxnet {
+namespace op {
+
+namespace {
+
+  template<typename DType, typename Rand>
+  void Shuffle1D(DType* out, index_t size, Rand* prnd) {
+    #ifdef USE_GNU_PARALLEL_SHUFFLE
+      auto rand_n = [prnd](index_t n) {
+        std::uniform_int_distribution<index_t> dist(0, n - 1);
+        return dist(*prnd);
+      };
+      __gnu_parallel::random_shuffle(out, out + size, rand_n);
+    #else
+      std::shuffle(out, out + size, *prnd);
+    #endif
+  }
+
+  template<typename DType, typename Rand>
+  void ShuffleND(DType* out, index_t size, index_t first_axis_len, Rand* prnd) {
+    // Fisher-Yates shuffling
+    const index_t stride = size / first_axis_len;
+    auto rand_n = [prnd](index_t n) {
+      std::uniform_int_distribution<index_t> dist(0, n - 1);
+      return dist(*prnd);
+    };
+    for (index_t i = first_axis_len - 1; i > 0; --i) {
```

Comment:
Add `CHECK_GT(first_axis_len, 0U);` above this line.

Reply:
Done.